workflows/vertex_classification/include/pando-wf1/gnn.hpp

```diff
@@ -0,0 +1,356 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2023. University of Texas at Austin. All rights reserved.
+
+#ifndef PANDO_WF1_GNN_HPP_
+#define PANDO_WF1_GNN_HPP_
+
+#include <utility>
+
+#include <pando-lib-galois/containers/dist_array.hpp>
+#include <pando-lib-galois/loops/do_all.hpp>
+
+#include <pando-rt/containers/vector.hpp>
+
+#include <pando-wf1/gnntypes.hpp>
+#include <pando-wf1/graphs/gnngraph.hpp>
+#include <pando-wf1/layers/fcn.hpp>
+#include <pando-wf1/layers/gcn.hpp>
+#include <pando-wf1/layers/layer.hpp>
+#include <pando-wf1/layers/softmax.hpp>
+#include <pando-wf1/optimizer.hpp>
+
+namespace gnn {
+
+/**
+ * @brief Class that manages overall graph neural network based training, testing, and validation.
+ *
+ * This class is the core of the graph neural network (GNN) workflow, and manages all training,
+ * testing, and validation phases. This class aggregates and orchestrates building blocks of GNN
+ * including neural network layers, activation layers, non-linear layers, an optimizer, and a graph.
+ *
+ * This class supports vertex classification, link prediction, and mulit-hop reasoning training.
+ *
+ */
+template <typename Graph>
+class GraphNeuralNetwork {
+public:
+  constexpr GraphNeuralNetwork() noexcept = default;
+
+  void initialize(pando::GlobalPtr<Graph> dGraphPtr) {
+    this->numPXNs = pando::getPlaceDims().node.id;
+    PANDO_CHECK(this->nodeList_.initialize(this->numPXNs));
+    for (std::int16_t n = 0; n < this->numPXNs; ++n) {
+      this->nodeList_[n] =
+          galois::PlaceType{pando::Place{pando::NodeIndex{n}, pando::anyPod, pando::anyCore},
+                            pando::MemoryType::Main};
+    }
+    this->gnnGraphPtr_ = static_cast<decltype(this->gnnGraphPtr_)>(
+        pando::getDefaultMainMemoryResource()->allocate(sizeof(graph::GNNGraph<Graph>)));
+    this->gnnGraphPtr_->initialize(dGraphPtr);
+    this->InitializeLayers(dGraphPtr);
+    // Setup optimizers
+    pando::Vector<LayerDimension> optimizerSizes;
+    PANDO_CHECK(optimizerSizes.initialize(0));
+    PANDO_CHECK(optimizerSizes.pushBack(this->gnnGraphPtr_->VertexFeatureLength() * 16));
+    PANDO_CHECK(optimizerSizes.pushBack(16 * this->gnnGraphPtr_->GetNumClasses()));
+    this->optimizer_.initialize(optimizerSizes, this->numGCNLayers_);
+  }
+
+  ~GraphNeuralNetwork() {
+    for (std::uint32_t l = 0; l < this->numGCNLayers_; ++l) {
+      pando::getDefaultMainMemoryResource()->deallocate(
+          static_cast<pando::GlobalPtr<GraphConvolutionalLayer<Graph>>>(this->gcnLayers_[l]),
+          sizeof(GraphConvolutionalLayer<Graph>));
+    }
+  }
+
+  /**
+   * @brief Allocates and initialize layers.
+   */
+  void InitializeLayers(pando::GlobalPtr<Graph> dGraphPtr) {
+    galois::HostIndexedMap<pando::Array<GNNFloat>> backwardOutputMatrix;
```

Comment:
These should be HostLocalStorage Right? because they are supposed to be located on each separate host.

Reply:
Let's table this till another PR.

Reply:
I think I can make this on the current PR. It seems like important and one of this PR's goals :-) I will do it asap. Thank you for letting me know.